src/lib.rs

```diff
@@ -201,13 +208,69 @@ impl_fft_with_opts_and_plan_for!(
     16
 );
 
-/// Utility function to separate interleaved format signals (i.e., Vector of Complex Number Structs)
-/// into separate vectors for the corresponding real and imaginary components.
-#[cfg(feature = "complex-nums")]
-pub fn separate_re_im<T: Float>(signal: &[Complex<T>]) -> (Vec<T>, Vec<T>) {
-    signal.iter().map(|z| (z.re, z.im)).unzip()
+macro_rules! impl_separate_re_im {
+    ($func_name:ident, $precision:ty, $lanes:literal, $simd_vec:ty) => {
+        /// Utility function to separate interleaved format signals (i.e., Vector of Complex Number Structs)
+        /// into separate vectors for the corresponding real and imaginary components.
+        #[multiversion::multiversion(
+                                    targets("x86_64+avx512f+avx512bw+avx512cd+avx512dq+avx512vl", // x86_64-v4
+                                            "x86_64+avx2+fma", // x86_64-v3
+                                            "x86_64+sse4.2", // x86_64-v2
+                                            "x86+avx512f+avx512bw+avx512cd+avx512dq+avx512vl",
+                                            "x86+avx2+fma",
+                                            "x86+sse4.2",
+                                            "x86+sse2",
+                                        ))]
+        pub fn $func_name(
+            signal: &[Complex<$precision>],
+        ) -> (Vec<$precision>, Vec<$precision>) {
+            let n = signal.len();
+            let mut reals = vec![0.0; n];
+            let mut imags = vec![0.0; n];
+
+            let complex_f64 = cast_slice(signal);
+            const CHUNK_SIZE: usize = $lanes * 2;
+
+            let mut i = 0;
```

Comment:
Why do we manually keep track of the index? I cannot see any places where it is actually used, is it an unused variable?

Reply:
It's used when taking care of the remainder

Reply:
Oh nevermind, I see it is used at the end. I am pretty sure we can infer it from the input length rather than keeping track of it in the hot loop. Something along the lines of `i = complex_f64.len() % CHUNK_SIZE` perhaps?

Reply:
@Shnatsel Let me know if the checked-in change suffices for this

Reply:
Looks good to me performance-wise, I defer to tests to verify correctness